distributed/shuffle/_worker_extension.py

```diff
@@ -185,15 +200,22 @@ def heartbeat(self) -> dict[str, Any]:
             "start": self.start_time,
         }
 
-    async def receive(self, data: list[bytes]) -> None:
+    async def receive(self, data: list[tuple[int, bytes]]) -> None:
         await self._receive(data)
 
-    async def _receive(self, data: list[bytes]) -> None:
+    async def _receive(self, data: list[tuple[int, bytes]]) -> None:
         self.raise_if_closed()
 
+        filtered = []
+        for d in data:
+            if d[0] not in self.received:
+                filtered.append(d[1])
+                self.received.add(d[0])
+                self.total_recvd += sizeof(d)
+        if not filtered:
+            return
         try:
-            self.total_recvd += sum(map(len, data))
-            groups = await self.offload(self._repartition_buffers, data)
+            groups = await self.offload(self._repartition_buffers, filtered)
```

Comment:
`del filtered` (wasn't there before, I know).

If pyarrow copies any data, which it has to in case of object types, for instance, an early del would free up data before everything is written to disk

Reply:
FWIW this is a bit more involved. if string objects are embedded in one of the byte objects, pyarrow is creating a copy of the str and holding the original in memory as well if some other, e.g. int array, is being referenced there. can't have it all